lib/cc/workspace/path_tree.rb

```diff
@@ -0,0 +1,115 @@
+require "pathname"
+require "set"
+
+module CC
+  class Workspace
+    class PathTree
+      def initialize(root_path, children = {})
+        @root_path = root_path.dup.freeze
+        @children = children
+      end
+
+      def clone
+        self.class.new(root_path, children.dup)
+      end
+
+      def exclude_paths(paths)
+        paths.each { |path| delete(normalized_path_pieces(path)) }
+      end
+
+      def include_paths(paths)
+        paths.each { |path| insert(normalized_path_pieces(path)) }
+      end
+
+      def all_paths
+        if populated?
+          children.values.flat_map(&:all_paths)
+        else
+          [File.join(root_path, File::SEPARATOR)]
+        end
+      end
+
+      protected
+
+      def populated?
+        children.present?
+      end
+
+      def delete(path_pieces)
+        populate_all
+
+        if 1 == path_pieces.size
+          children.delete(path_pieces[0])
+        else
+          child = children[path_pieces[0]]
+          if child
+            child.delete(path_pieces.drop(1))
+            children.delete(path_pieces[0]) unless child.populated?
+          else
+            CLI.debug(
+              "Couldn't exclude because part of path doesn't exist or was already excluded.",
```

Comment:
I can be convinced that we should keep the logging in `#insert` (and maybe make it a WARN even). But for excludes, I think we should silently ignore.

Two reasons:

- When I create a git project, I start with a `.gitignore` full of things that don't exist (yet or never will), this is super common and I see analysis exclusions the same way
- Something not existing clearly won't be analyzed, i.e. it's been excluded -- technically speaking, we've done the Right Thing